File: src/event/ResourceMetaUpdated.php
<?php

namespace Ething\Event;

use \Ething\ResourceQueryParser;


class ResourceMetaUpdated extends AbstractResourceEvent {
	
	
	
	static public function emit(\Ething\Resource $resource, array $attributes){
		return new Signal('ResourceMetaUpdated', array(
			'resource' => $resource->id(), // the updated resource
			'resourceName' => $resource->name(),
			'attributes' => $attributes // list of names of the attributes that has been updated
		));
	}
	
	
	static public function validate(array &$attributes, array $context){
		
		$context['keyValidator'] = function($key, &$value,$context){
			
			switch($key){
				
				case 'attributes': // list of names of the attributes to listen to some update
					
					if(\Ething\Helpers::is_numeric_array($value)){
						foreach($value as $attr){
							if(!is_string($attr))
								throw new \Exception("the '{$key}' must be an array of string");
						}
					} else if(!is_null($value))
						throw new \Exception("the '{$key}' must be an array of string");
					
					break;
					
				case 'filter': // an expression to use as an extended filter
					
					if(is_string($value)){
						// parse the expression
						if(!ResourceQueryParser::check($context['ething'], $value, $error))
							throw new \Exception('invalid expression: '.$error);
					} else if(!is_null($value))
						throw new \Exception("the '{$key}' must be a string");
					
					break;
				
				default:
					return false;
			}
			
		};
		
		return parent::validate($attributes, $context);
	}
	
	protected function call(Signal $signal){
		if((!isset($this->attributes) || count(array_intersect($this->attributes, $signal->attributes))) && parent::call($signal)){
			return !isset($this->filter) || boolval( $this->ething()->findOne( array(
					'$and' => array(
						array( '_id' => $signal->resource ),
						$this->ething()->resourceQueryParser()->parse($this->filter)
					)
				) ) );
		}
		return false;
	}
	
}

File: ething/rule/event/ResourceMetaUpdated.py
# coding: utf-8
from future.utils import string_types

from .ResourceEvent import ResourceEvent


class ResourceMetaUpdated(ResourceEvent):
    
    
    @classmethod
    def emit(cls, resource, attributes):
        return super(ResourceMetaUpdated, cls).emit(resource, {
            'attributes': attributes # list of names of the attributes that has been updated
        })
    
    @staticmethod
    def validate (attributes, context):
        
        attributes.setdefault('attributes', None)
        
        other = {}
        
        for key in attributes:
            
            value = attributes[key]
            
            if key == 'attributes':
                
                if value is None:
                    pass
                elif isinstance(value, list):
                    for v in value:
                        if not isinstance(v, string_types):
                            raise Exception("the key '%s' must be an array of string" % key)
                else:
                    raise Exception("the key '%s' must be an array of string" % key)
                
            else:
                other[key] = value
        
        return ResourceEvent.validate(other, context)
    
    
    def call(self, signal):
        
        if self['attributes'] is not None:
            
            a = self['attributes']
            b = signal['attributes']
            f = False
            for val in a:
                if val in b:
                    f = True
                    break
            if not f:
                return False
        
        return super(ResourceMetaUpdated, self).call(signal)
    


